File: head_to_offer/14/hto_14.java
import java.util.Arrays;
import java.util.Comparator;
import java.util.Deque;
import java.util.ArrayDeque;
import java.util.List;
import java.util.ArrayList;

class Solution {
    public int cuttingRope(int n) {
        if (n == 2) return 1;
        if (n == 3) return 2;
        int[] res = new int[n+1];
        res[1] = 1;
        res[2] = 2;
        res[3] = 3;
        for (int i = 4; i <= n; ++i) {
            for (int j = 1; j <= i / 2; ++j)
                res[i] = Math.max(res[i], res[j] * res[i - j]);
        }
        return res[n];
    }
}

public class leetcode {
    public static void main(String[] args) {
        Solution s = new Solution();
        System.out.println(s.cuttingRope(10));
    }
}

File: head_to_offer/14/hto_14.c
#include <stdio.h>
#include <math.h>

int cuttingRope(int n) {
    if (n == 2) return 1;
    if (n == 3) return 2;
    int a = 1;
    while (n > 4) {
        a *= 3;
        n -= 3;
    }
    return a * n;
}

int cuttingRope1(int n) {
    if (n <= 3) return n - 1;
    int a = n / 3, b = n % 3;
    if (b == 0) return pow(3.0, a);
    if (b == 1) return pow(3.0, a - 1) * 4;
    return pow(3.0, a) * 2;
}

// compile add "-lm"
int main(void) {
    printf("%d\n", cuttingRope(2));
}